### PR TITLE
[SPARK-21351][SQL] Remove the UpdateAttributeNullability rule from the optimizer

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr removed `UpdateAttributeNullability` from the optimizer because the same logic happens in the analyzer. See SPARK-26459(#23390) for more detailed discussion.

## How was this patch tested?
N/A

### CODE REVIEW

**Comment by maropu** [comment]
@cloud-fan @gatorsmile It seems the removal don't affect the existing tests now.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UpdateAttributeNullabilityInOptimizerSuite.scala [hunk 1]
```diff
@@ -1,58 +0,0 @@
...
-
-  test("update nullability in AttributeReference")  {
-    val rel = LocalRelation('a.long.notNull)
-    // In the 'original' plans below, the Aggregate node produced by groupBy() has a
-    // nullable AttributeReference to `b`, because both array indexing and map lookup are
-    // nullable expressions. After optimization, the same attribute is now non-nullable,
-    // but the AttributeReference is not updated to reflect this. So, we need to update nullability
-    // by the `UpdateAttributeNullability` rule.
-    val original = rel
-      .select(GetArrayItem(CreateArray(Seq('a, 'a + 1L)), 0) as "b")
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UpdateAttributeNullabilityInOptimizerSuite.scala]: Shall we update `GetArrayItem.nullable`, to special-handle the combination of `CreateArray` and `Literal`?

Comment (maropu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UpdateAttributeNullabilityInOptimizerSuite.scala]: Ah, I see, we can do so. Is it ok to file a separate jira and make a pr for that in following activities?

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UpdateAttributeNullabilityInOptimizerSuite.scala]: sure

Comment (maropu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UpdateAttributeNullabilityInOptimizerSuite.scala]: Made a pr in #23566

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by maropu** [comment]
ping @cloud-fan @gatorsmile @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
Hi, @maropu . 
Based on the our discussion history and the recent nullability fixes, we decided to remove this from the optimizer step.

As a the final piece of this removal, could you remove the following `UpdateAttributeNullability` optimizer description? Otherwise, someone may try to add it back.
- https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UpdateAttributeNullability.scala#L32-L34
```
 *
 * This rule should be executed again at the end of optimization phase, as optimizer may change
 * some expressions and their nullabilities as well. See SPARK-21351 for more details.
```

**Comment by maropu** [comment]
oh, yeah. I'll do that.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**